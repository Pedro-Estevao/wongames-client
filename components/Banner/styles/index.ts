'use client';

import styled, { css } from 'styled-components';

type ImageProps = {
	src: string;
};

export const BannerContainer = styled.div``;

export const Image = styled.div<ImageProps>`
	${({ src }) => css`
		background-image: url(${src});
		background-position: center center;
		background-size: cover;
	`}
`;
