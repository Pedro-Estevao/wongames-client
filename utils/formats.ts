export const formatPrice = (price: number | bigint) => {
	return new Intl.NumberFormat('en', {
		style: 'currency',
		currency: 'USD'
	}).format(price);
};

export const formatDate = (date: string) => {
	return new Intl.DateTimeFormat('en-US', {
		day: 'numeric',
		month: 'short',
		year: 'numeric'
	}).format(new Date(date));
};

export const formatImageUrl = (url: string | undefined) => {
	if (process.env.NEXT_PUBLIC_API_URL) {
		return `${process.env.NEXT_PUBLIC_API_URL}${url}`;
	}

	if (url) {
		return url;
	}

	return null;
};