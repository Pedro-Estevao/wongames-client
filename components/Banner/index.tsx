import React from 'react';
import * as S from './styles';
import Button from '../Button';
import Ribbon, { RibbonColorsProps, RibbonSizeProps } from '../Ribbon';
import Link from 'next/link';
import Image from "next/image";

export type BannerProps = {
	img: string;
	title: string;
	subTitle: string;
	buttonLabel: string;
	buttonLink: string;
	ribbon?: React.ReactNode;
	ribbonColor?: RibbonColorsProps;
	ribbonSize?: RibbonSizeProps;
};

const Banner = ({
	img,
	title,
	subTitle,
	buttonLabel,
	buttonLink,
	ribbon,
	ribbonColor = 'primary',
	ribbonSize = 'normal'
}: BannerProps) => {
	return (
		<S.BannerContainer>
			{!!ribbon && (
				<Ribbon color={ribbonColor} size={ribbonSize}>
					{ribbon}
				</Ribbon>
			)}

			<S.ImageWrapper>
				<Image src={img} fill className="object-cover" alt={title} />
			</S.ImageWrapper>

			<S.Caption>
				<S.Title>{title}</S.Title>
				<S.Subtitle dangerouslySetInnerHTML={{ __html: subTitle }}  />
				<Button as={Link} href={buttonLink} size="large">
					{buttonLabel}
				</Button>
			</S.Caption>
		</S.BannerContainer>
	);
};

export default Banner;
