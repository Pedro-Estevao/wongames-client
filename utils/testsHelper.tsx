import { CartContext, CartContextData, CartContextDefaultValues } from "@/hooks/useCart";
import { WishlistContext, WishlistContextData, WishlistContextDefaultValues } from "@/hooks/useWishlist";
import { theme } from "@/public/styles/theme";
import { render, RenderOptions } from "@testing-library/react";
import { ThemeProvider } from "styled-components";

type CustomRenderProps = {
    cartProviderProps?: CartContextData;
    wishlistProviderProps?: WishlistContextData;
} & Omit<RenderOptions, 'queries'>;

const customRender = (
    ui: React.ReactElement,
    {
        cartProviderProps = CartContextDefaultValues,
        wishlistProviderProps = WishlistContextDefaultValues,
        ...renderOptions
    }: CustomRenderProps = {}
) => render(
    <ThemeProvider theme={theme}>
        <CartContext.Provider value={cartProviderProps}>
            <WishlistContext.Provider value={wishlistProviderProps}>
                {ui}
            </WishlistContext.Provider>
        </CartContext.Provider>
    </ThemeProvider>,
    renderOptions
);

export * from "@testing-library/react";
export { customRender as render };