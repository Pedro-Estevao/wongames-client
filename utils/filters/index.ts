import { ItemProps } from '@/components/ExploreSidebar';
import { FilterObject, FilterValue } from "@/components/Layouts/Games";
import { ParsedUrlQueryInput } from 'querystring';

type ParseArgs = {
	queryString: ParsedUrlQueryInput;
	filterItems: Pick<ItemProps, 'type' | 'name'>[];
};

export const parseQueryStringToFilter = ({
	queryString,
	filterItems
}: ParseArgs) => {
	const obj: Record<string, string | string[]> = {};

	Object.keys(queryString).forEach((key) => {
		const item = filterItems?.find((item) => item.name === key);
		const isCheckbox = item?.type === 'checkbox';
		const queryValue = queryString[key];

		if (Array.isArray(queryValue)) {
			// Remove duplicatas do array
			obj[key] = Array.from(new Set(queryValue));
		} else if (typeof queryValue === 'string') {
			if (isCheckbox) {
				// Se for um checkbox, converte a string em array e remove duplicatas
				const values = queryValue.split(',').filter(Boolean); // Remove strings vazias
				obj[key] = Array.from(new Set(values));
			} else {
				// Se não for checkbox, mantém a string como está
				obj[key] = queryValue;
			}
		}
	});

	return obj;
};

export const parseQueryStringToObjFilter = (
	queryString: Record<string, string | string[]>,
	filterItems: ItemProps[]
): FilterObject => {
	const filters: FilterObject = {};

	filterItems.filter((item) => item.name !== 'sort').forEach((item) => {
		if (queryString[item.name]) {
			const queryValue = queryString[item.name];
			let filterValue: FilterValue | string | string[] | number;

			if (Array.isArray(queryValue)) {
				filterValue = queryValue;
			} else {
				// Check if the value is a number (used for price)
				filterValue = isNaN(Number(queryValue)) ? queryValue : Number(queryValue);

				if (item.type === 'checkbox' && typeof filterValue === 'string') {
					filterValue = filterValue.split(',').map(str => str.trim());
				}
			}

			filters[item.name] = item.value
				? item.filter
					? { [item.value]: { [item.filter]: filterValue } }
					: { [item.value]: filterValue }
				: item.filter
					? { [item.filter]: filterValue }
					: filterValue;
		}
	});

	return filters;
};