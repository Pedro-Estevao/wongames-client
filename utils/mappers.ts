import { BannerProps } from '@/components/Banner';
import { GameCardProps } from '@/components/GameCard';
import { HighlightProps } from '@/components/Highlight';
import {
	BannerEntity,
	BannerEntityResponseCollection,
	BannerFragmentFragment,
	GameEntityResponseCollection,
	GameFragmentFragment,
	HighlightFragmentFragment
} from '@/graphql/generated/graphql';

export const bannerMapper = (banners: BannerEntityResponseCollection) => {
	return (
		(banners?.data?.map((banner: BannerEntity) => ({
			img: `${process.env.NEXT_PUBLIC_API_URL}${(
				banner.attributes as BannerFragmentFragment
			)?.image?.data?.attributes?.url}`,
			title: (banner.attributes as BannerFragmentFragment)?.title || '',
			subTitle: (banner.attributes as BannerFragmentFragment)?.subTitle || '',
			buttonLabel:
				(banner.attributes as BannerFragmentFragment)?.button?.label || '',
			buttonLink:
				(banner.attributes as BannerFragmentFragment)?.button?.link || '',
			...((banner.attributes as BannerFragmentFragment)?.ribbon && {
				ribbon:
					(banner.attributes as BannerFragmentFragment)?.ribbon?.text || null,
				ribbonColor:
					(banner.attributes as BannerFragmentFragment)?.ribbon?.color || null,
				ribbonSize:
					(banner.attributes as BannerFragmentFragment)?.ribbon?.size || null
			})
		})) as BannerProps[]) || []
	);
};

export const gamesMapper = (games: GameEntityResponseCollection) => {
	return (
		(games?.data.map((game) => ({
			title: (game.attributes as GameFragmentFragment).name,
			slug: (game.attributes as GameFragmentFragment).slug,
			developer: (game.attributes as GameFragmentFragment).developers?.data[0]
				.attributes?.name,
			img: `${process.env.NEXT_PUBLIC_API_URL}${(game.attributes as GameFragmentFragment)
				.cover?.data?.attributes?.url}`,
			price: (game.attributes as GameFragmentFragment).price
		})) as GameCardProps[]) || []
	);
};

export const staticGamesParams = (games: GameEntityResponseCollection) => {
	return (
		(games?.data.map((game) => ({
			title: (game.attributes as GameFragmentFragment).name,
			slug: (game.attributes as GameFragmentFragment).slug,
			developer: (game.attributes as GameFragmentFragment).developers?.data[0]
				.attributes?.name,
			img: `${process.env.NEXT_PUBLIC_API_URL}${(game.attributes as GameFragmentFragment)
				.cover?.data?.attributes?.url}`,
			price: (game.attributes as GameFragmentFragment).price
		})) as GameCardProps[]) || []
	);
};

export const highlightMapper = (highlight: HighlightFragmentFragment) => {
	return (
		highlight &&
		({
			title: highlight.title,
			subTitle: highlight.subTitle,
			backgroundImage: `${process.env.NEXT_PUBLIC_API_URL}${highlight.background?.data?.attributes?.url}`,
			floatImage: `${process.env.NEXT_PUBLIC_API_URL}${highlight.floatImage?.data?.attributes?.url}`,
			buttonLabel: highlight.buttonLabel,
			buttonLink: highlight.buttonLink,
			alignment: highlight.alignment
		} as HighlightProps)
	);
};
