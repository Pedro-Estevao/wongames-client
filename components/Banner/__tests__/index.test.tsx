import { render, screen } from '@/utils/testsHelper';
import { expect } from '@jest/globals';
import Banner from "@/components/Banner";

const props = {
    img: 'https://images.unsplash.com/photo-1674461315270-d6b04eaff631',
    title: 'Defy death',
    subTitle: '<p>Play the new <strong>CrashLands</strong> season</p>',
    buttonLabel: 'Buy now',
    buttonLink: '/games/defy-death',
};

describe('<Banner />', () => {
    it('should render correctly', () => {
        const { container } = render(<Banner {...props} />);
        const { firstChild } = container;

        expect(screen.getByRole('heading', { name: /Defy death/i })).toBeInTheDocument();

        expect(screen.getByRole('heading', { name: /Play the new CrashLands season/i })).toBeInTheDocument();

        expect(screen.getByRole('img', { name: /Defy death/i })).toBeInTheDocument();

        expect(firstChild).toMatchSnapshot();
    });

    it('should render a Ribbon', () => {
        render(
            <Banner
                {...props}
                ribbon="20% OFF"
                ribbonSize="small"
                ribbonColor="secondary"
            />
        );

        const ribbon = screen.getByText(/20% OFF/i);

        expect(ribbon).toBeInTheDocument();
        expect(ribbon).toHaveStyle({ backgroundColor: '#3CD3C1' });
        expect(ribbon).toHaveStyle({ height: '2.6rem', fontSize: '1.2rem' });
    });
});