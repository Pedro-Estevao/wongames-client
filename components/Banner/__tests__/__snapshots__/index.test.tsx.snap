// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Banner /> should render correctly 1`] = `
.c0 {
  position: relative;
}

.c1 {
  width: 100%;
  height: 23rem;
  background-color: #EAEAEA;
  position: relative;
}

.c2 {
  width: 100%;
  background-color: rgba(0,0,0,0.7);
  padding: 2.4rem;
}

.c3 {
  font-size: 1.8rem;
  font-weight: 600;
  color: #FAFAFA;
}

.c4 {
  color: #FAFAFA;
  font-size: 1.4rem;
  font-weight: 400;
  margin-bottom: 1.6rem;
}

.c4 strong {
  color: #F231A5;
  font-weight: 600;
}

.c5 {
  display: inline-flex;
  align-items: center;
  justify-content: center;
  background: linear-gradient(180deg,#ff5f5f 0%,#f062c0 50%);
  color: #FAFAFA;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  border: 0;
  cursor: pointer;
  border-radius: 0.4rem;
  padding: 0.8rem;
  text-decoration: none;
  height: 5rem;
  font-size: 1.6rem;
  padding: 0.8rem 4.8rem;
}

.c5:focus {
  box-shadow: 0 0 0 3px #3CD3C1;
}

.c5:hover {
  background: linear-gradient(180deg, #e35565 0%, #d958a6 50%);
}

@media (max-width: 1170px) {

}

@media (min-width: 768px) {
  .c0 {
    box-shadow: 0 0.4rem 0.5rem 0 rgba(0, 0, 0, 0.2);
  }
}

@media (min-width: 768px) {
  .c1 {
    height: 58rem;
  }
}

@media (min-width: 768px) {
  .c2 {
    border-radius: 0 0 0.4rem 0.4rem;
    padding: 4.0rem;
    position: absolute;
    bottom: 0;
    left: 0;
  }
}

@media (min-width: 768px) {
  .c3 {
    font-size: 2.8rem;
  }
}

@media (min-width: 768px) {
  .c4 {
    font-size: 1.8rem;
  }
}

<main
  class="c0"
>
  <div
    class="c1"
  >
    <img
      alt="Defy death"
      class="object-cover"
      data-nimg="fill"
      decoding="async"
      loading="lazy"
      sizes="100vw"
      src="/_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1674461315270-d6b04eaff631&w=3840&q=75"
      srcset="/_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1674461315270-d6b04eaff631&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1674461315270-d6b04eaff631&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1674461315270-d6b04eaff631&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1674461315270-d6b04eaff631&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1674461315270-d6b04eaff631&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1674461315270-d6b04eaff631&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1674461315270-d6b04eaff631&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1674461315270-d6b04eaff631&w=3840&q=75 3840w"
      style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
    />
  </div>
  <div
    class="c2"
  >
    <h2
      class="c3"
    >
      Defy death
    </h2>
    <h3
      class="c4"
    >
      <p>
        Play the new 
        <strong>
          CrashLands
        </strong>
         season
      </p>
    </h3>
    <a
      class="c5"
      href="/games/defy-death"
    >
      <span>
        Buy now
      </span>
    </a>
  </div>
</main>
`;
